packages/core/src/browser/shell/dirty-widget-guard.ts

```diff
@@ -0,0 +1,33 @@
+/********************************************************************************
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { ApplicationShell } from './application-shell';
+import { FrontendApplicationContribution } from '../frontend-application';
+
+@injectable()
+export class DirtyWidgetGuard implements FrontendApplicationContribution {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    preventStop() {
+        if (this.shell.canSaveAll()) {
+            return 'Some documents should be saved, data will be lost otherwise.';
```

Comment:
How about checking preference and displaying errors to a user here, not into `FrontendApplication`?